torch/_refs/__init__.py

```diff
@@ -1982,6 +1984,75 @@ def conj(input: TensorLikeType) -> TensorLikeType:
     return prims.conj(input)
 
 
+def constant_pad_nd(
+    input: TensorLikeType, pad: List[int], value: NumberType = 0
+) -> TensorLikeType:
+    if len(pad) % 2 != 0:
+        raise RuntimeError(
+            f"Length of pad must be even but instead it equals {len(pad)}"
+        )
+
+    input_sizes = input.shape
+    l_inp = len(input_sizes)
+
+    l_pad = len(pad) // 2
+    l_diff = l_inp - l_pad
+
+    if l_inp < l_pad:
+        raise RuntimeError(
+            "Length of pad should be no more than twice the number of "
+            f"dimensions of the input. Pad length is {len(pad)} while the input has ",
+            f"{l_inp} dimensions.",
+        )
+
+    c_input = input
+    for i in range(l_diff, l_inp):
+        pad_idx = 2 * (l_inp - i - 1)
+        if pad[pad_idx] < 0:
+            c_input = narrow(c_input, i, -pad[pad_idx], c_input.shape[i] + pad[pad_idx])
+
+        if pad[pad_idx + 1] < 0:
+            c_input = narrow(c_input, i, 0, c_input.shape[i] + pad[pad_idx + 1])
+
+    # if none of the pads are positive we can just return the result
+    if builtins.all(p <= 0 for p in pad):
+        return prims.clone(c_input)
+
+    new_shape = list(input_sizes[:l_diff])
+
+    for i in range(l_pad):
+        pad_idx = len(pad) - ((i + 1) * 2)
+        new_dim = input_sizes[l_diff + i] + pad[pad_idx] + pad[pad_idx + 1]
+        if new_dim <= 0:
+            raise RuntimeError(
+                f"The input size {input_sizes[l_diff + i]}, plus negative padding ",
+                f"{pad[pad_idx]} and {pad[pad_idx + 1]} resulted in a negative output size, "
+                f"which is invalid. Check dimension {l_diff + i} of your input.",
+            )
+        new_shape.append(new_dim)
+
+    options = dict(
+        dtype=input.dtype, device=input.device, requires_grad=input.requires_grad
+    )
+    if value == 0:
+        output = zeros(new_shape, **options)
+    else:
+        output = full(new_shape, value, **options)
+
+    c_output = output
+    for i in range(l_diff, l_inp):
+        pad_idx = 2 * (l_inp - i - 1)
+        if pad[pad_idx] > 0:
+            c_output = narrow(
```

Comment:
should this be `torch.narrow` instead of `narrow`? The general rule is to use `torch` calls?

Reply:
I neglected to mention this. Prefer using `torch.foo` unless `refs.foo` has some different and necessary functionality. The call to `torch.foo` is remapped to `refs.foo` under a context manager in some tests

Reply:
Is the reasoning for this documented somewhere? It seems counter-intuitive to me.

Reply:
You're correct and refs do go to prims, but in real PyTorch programs we'll have to remap ops like `torch.foo` to `refs.foo` to decompose them to prims

Your proposal is reasonable, @IvanYashchuk, but every ref should be registered as a decomposition for the corresponding operator, and in that case we can still generally prefer calling ops in the `torch` namespace